cdap-app-templates/cdap-etl/cdap-data-streams-base/src/main/java/io/cdap/cdap/datastreams/DataStreamsPipelineSpecGenerator.java

```diff
@@ -79,22 +93,79 @@ public DataStreamsPipelineSpec generateSpec(DataStreamsConfig config) throws Val
     DataStreamsPipelineSpec.Builder specBuilder = DataStreamsPipelineSpec.builder(batchIntervalMillis, pipelineId)
       .setExtraJavaOpts(config.getExtraJavaOpts())
       .setStopGracefully(config.getStopGracefully())
-      .setIsUnitTest(config.isUnitTest())
-      .setCheckpointsDisabled(config.checkpointsDisabled());
-    String checkpointDir = config.getCheckpointDir();
-    if (!config.checkpointsDisabled() && checkpointDir != null) {
-      try {
-        new Path(checkpointDir);
-      } catch (Exception e) {
-        throw new IllegalArgumentException(
-          String.format("Checkpoint directory '%s' is not a valid Path: %s", checkpointDir, e.getMessage()), e);
-      }
-      specBuilder.setCheckpointDirectory(checkpointDir);
-    }
+      .setIsUnitTest(config.isUnitTest());
+
     configureStages(config, specBuilder);
+    //Configure the at least once processing mode for this pipeline
+    configureAtleastOnceMode(config, specBuilder);
     return specBuilder.build();
   }
 
+  @VisibleForTesting
+  void configureAtleastOnceMode(DataStreamsConfig config, DataStreamsPipelineSpec.Builder specBuilder) {
+    //If runtime arg sets atleast once processing to false, disable both
+    if (runtimeConfigurer != null && runtimeConfigurer.getRuntimeArguments() != null) {
+      boolean atleastOnceProcessingEnabled = Boolean.parseBoolean(
+        runtimeConfigurer.getRuntimeArguments().getOrDefault(CDAP_STREAMING_ATLEASTONCE_ENABLED, "true"));
+      if (!atleastOnceProcessingEnabled) {
+        specBuilder.setNativeStateTrackingEnabled(false);
+        specBuilder.setCheckpointsDisabled(true);
+        specBuilder.setCheckpointDirectory(null);
+        return;
+      }
+    }
+    //Check if native state tracking is possible
+    if (nativeStateTrackingSupported(config, sourcePluginTypes, stateHandlingSources)) {
+      // Native state tracking and spark checkpointing is mutually exclusive
+      // This is because Spark recreates context from checkpoint data
+      specBuilder.setNativeStateTrackingEnabled(true);
+      specBuilder.setCheckpointsDisabled(true);
+      specBuilder.setCheckpointDirectory(null);
+    } else {
+      specBuilder.setCheckpointsDisabled(false);
```

Comment:
Add a comment here sayint that in this case, checkpointing is enabled

Reply:
With renaming of the variable, this should be evident now.